crypto/rand/drbg_lib.c

```diff
@@ -366,6 +376,15 @@ int RAND_DRBG_generate(RAND_DRBG *drbg, unsigned char *out, size_t outlen,
     if (drbg->reseed_counter >= drbg->reseed_interval)
         drbg->state = DRBG_RESEED;
 
+    if (drbg->parent != NULL) {
+        if (drbg->parent->lock)
+            CRYPTO_THREAD_read_lock(drbg->parent->lock);
+        if (drbg->restart_count != drbg->parent->restart_count)
```

Comment:
Is the locking needed here?  I thought integer reads were considered atomic.

It might require an atomic increment when the count is changed but that's cheaper than locking when it is available.

Reply:
Then I just leave it as it is: Only the three global DRBGs are created with a lock, the other DRBGs don’t have a lock by default. For completeness sake, I would add a RAND_DRBG_enable_locking() method to the RAND_DRBG API.

I agree with you that letting `CRYPTO_THREAD_xxx_lock()` accept NULL pointers as no-ops is probably not a good idea. So the call needs to be wrapped by a conditional.

Reply:
Thanks for suggesting the atomic operations.  I hope the following question does not sound naive: I was wondering whether any special precaution is necessary in the first place. Even in the unlikely case that due to concurrency issues the if-block

```
if (drbg->restart_count != drbg->parent->restart_count) {
   …
}
```
is not entered erroneously, this would only mean that automatic restarting of the chained DRBG would be postponed until the next generate request. And failing twice in a row is extremely unlikely. What do you think about this kind of reasoning?

Reply:
I would prefer the CRYPTO_atomic_read(), not for performance reasons but for clarity. Provided that's not too much effort for you to add.

Reply:
CRYPTO_atomic_read sounds like a good idea.

Reply:
If there is a race reading either `drbg->restart_count` or `drbg->parent->restart_count`, a reseed will occur now and again the next time the DRBG is used.  I.e. not a big concern.  If a stale value remains in cache, the reseed will be delayed.

Reply:
See #4414